doc/reference/devicetree/index.rst

```diff
@@ -9,46 +9,111 @@ scheduling. They can be used from any calling context and at file scope.
 before they can be used; these are discussed individually below. These macros
-are generally meant for use within device drivers.
+are generally meant for use within :ref:`device drivers <device_model_api>`,
+though they can be used outside of drivers with appropriate care.
 
 .. _devicetree-generic-apis:
 
 Generic APIs
 ************
 
-These APIs can be used anywhere.
+The APIs in this section can be used anywhere and do not require
+``DT_DRV_COMPAT`` to be defined.
 
 Node identifiers
 ================
 
-You can use node identifiers for devicetree nodes which are enabled (i.e. have
-``status = "okay";`` properties) and have matching compatibles. This can be
-tested with :c:func:`DT_NODE_HAS_STATUS`.
+A *node identifier* is a way to refer to a devicetree node at C preprocessor
+time. While node identifiers are not C values, you can use them to access
+devicetree data in C rvalue form using, for example, the
+:ref:`devicetree-property-access` API.
+
+The root node ``/`` has node identifier ``DT_ROOT``. You can create node
+identifiers for other devicetree nodes using :c:func:`DT_PATH`,
+:c:func:`DT_NODELABEL`, :c:func:`DT_ALIAS`, and :c:func:`DT_INST`.
+
+There are also :c:func:`DT_PARENT` and :c:func:`DT_CHILD` macros which can be
+used to create node identifiers for a given node's parent node or a particular
+child node, respectively.
 
 .. doxygengroup:: devicetree-generic-id
    :project: Zephyr
 
+.. _devicetree-property-access:
+
 Property access
 ===============
 
+The following general-purpose macros can be used to access node properties.
+There are special-purpose APIs for accessing the :ref:`devicetree-reg-property`
+and :ref:`devicetree-interrupts-property`.
+
+Property values can be read using these macros even if the node is disabled,
+as long as it has a matching binding.
+
 .. doxygengroup:: devicetree-generic-prop
    :project: Zephyr
 
-Chosen nodes
-============
+.. _devicetree-reg-property:
 
-.. doxygengroup:: devicetree-generic-chosen
+``reg`` property
+================
+
+Use these APIs instead of :ref:`devicetree-property-access` to access the
+``reg`` property. Because this property's semantics are defined by the
+devicetree specification, these macros can be used even for nodes without
+matching bindings.
+
+.. doxygengroup:: devicetree-reg-prop
+   :project: Zephyr
+
+.. _devicetree-interrupts-property:
+
+``interrupts`` property
+=======================
+
+Use these APIs instead of :ref:`devicetree-property-access` to access the
+``interrupts`` property.
+
+Because this property's semantics are defined by the devicetree specification,
+some of these macros can be used even for nodes without matching bindings. This
+does not apply to macros which take cell names as arguments.
+
+.. doxygengroup:: devicetree-interrupts-prop
+   :project: Zephyr
+
+For-each macros
+===============
+
+There is currently only one generic for-each macro, :c:func:`DT_FOREACH_CHILD`,
```

Comment:
This is a little confusing.  We only have DT_FOREACH_CHILD, but then the next paragraph references DT_INST_FOREACH_STATUS_OKAY.  So saying 'only one' might be seen as confusing.

Reply:
This is the only "generic" one.

Reply:
Changed the wording to hopefully make this more clear.